File: trifecta-ios/Trifecta/App/AppEnvironment.swift
import Foundation
import SwiftUI

struct SavedServerProfile: Codable, Identifiable, Equatable {
    let id: String
    var name: String
    var serverURL: URL
    var createdAt: Date
    var lastUsedAt: Date
}

@Observable
final class AppEnvironment {
    enum SessionState: Equatable {
        case unconfigured
        case configured(URL)
    }

    var sessionState: SessionState
    var connectionStatus: ConnectionState = .offline
    var threadList: ThreadListStore
    var savedProfiles: [SavedServerProfile] = []
    var activeProfileID: SavedServerProfile.ID?
    var serverConfig: ServerRuntimeConfig?
    var serverConfigError: String?
    private(set) var connection: T3Connection?
    private(set) var client: T3Client?
    private var statusTask: Task<Void, Never>?
    private let profilesDefaultsKey = "savedServerProfiles"
    private let activeProfileDefaultsKey = "activeServerProfileID"
    private let tokenPrefix = "t3.bearer.token.profile."

    init() {
        let bootstrap = Self.bootstrapState()
        self.threadList = ThreadListStore()
        self.savedProfiles = bootstrap.savedProfiles
        self.activeProfileID = bootstrap.activeProfileID
        self.sessionState = bootstrap.sessionState
    }

    func configure(serverURL: URL, bearerToken: String, name: String? = nil) async {
        let profileName = name?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
            ? name!.trimmingCharacters(in: .whitespacesAndNewlines)
            : (serverURL.host ?? serverURL.absoluteString)
        let profileID = upsertProfile(serverURL: serverURL, name: profileName)
        KeychainStore.save(bearerToken, account: tokenAccount(for: profileID))
        activeProfileID = profileID
        UserDefaults.standard.set(profileID, forKey: activeProfileDefaultsKey)
        sessionState = .configured(serverURL)
        await startClient(serverURL: serverURL, bearerToken: bearerToken)
    }

    func resumeIfConfigured() async {
        guard let profile = activeProfile ?? savedProfiles.sorted(by: { $0.lastUsedAt > $1.lastUsedAt }).first,
              let token = KeychainStore.read(account: tokenAccount(for: profile.id)) else {
            return
        }
        activeProfileID = profile.id
        UserDefaults.standard.set(profile.id, forKey: activeProfileDefaultsKey)
        markProfileUsed(profile.id)
        sessionState = .configured(profile.serverURL)
        await startClient(serverURL: profile.serverURL, bearerToken: token)
    }

    func switchToProfile(id: SavedServerProfile.ID) async {
        guard let profile = savedProfiles.first(where: { $0.id == id }),
              let token = KeychainStore.read(account: tokenAccount(for: id)) else {
            return
        }
        activeProfileID = id
        UserDefaults.standard.set(id, forKey: activeProfileDefaultsKey)
        markProfileUsed(id)
        sessionState = .configured(profile.serverURL)
        await startClient(serverURL: profile.serverURL, bearerToken: token)
    }

    func renameProfile(id: SavedServerProfile.ID, name: String) {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, let index = savedProfiles.firstIndex(where: { $0.id == id }) else {
            return
        }
        savedProfiles[index].name = trimmed
        persistProfiles()
    }

    func removeProfile(id: SavedServerProfile.ID) async {
        let wasActive = (activeProfileID == id)
        KeychainStore.delete(account: tokenAccount(for: id))
        savedProfiles.removeAll { $0.id == id }
        persistProfiles()

        guard wasActive else { return }

        if let client { await client.stop() }
        client = nil
        connection = nil
        threadList = ThreadListStore()
        serverConfig = nil
        serverConfigError = nil
        connectionStatus = .offline

        if let next = savedProfiles.sorted(by: { $0.lastUsedAt > $1.lastUsedAt }).first {
            activeProfileID = next.id
            UserDefaults.standard.set(next.id, forKey: activeProfileDefaultsKey)
            sessionState = .configured(next.serverURL)
            if let token = KeychainStore.read(account: tokenAccount(for: next.id)) {
                await startClient(serverURL: next.serverURL, bearerToken: token)
            }
        } else {
            activeProfileID = nil
            UserDefaults.standard.removeObject(forKey: activeProfileDefaultsKey)
            sessionState = .unconfigured
        }
    }

    func signOut() async {
        statusTask?.cancel()
        statusTask = nil
        if let client { await client.stop() }
        client = nil
        connection = nil
        if let activeProfileID {
            KeychainStore.delete(account: tokenAccount(for: activeProfileID))
            savedProfiles.removeAll { $0.id == activeProfileID }
            persistProfiles()
            self.activeProfileID = nil
            UserDefaults.standard.removeObject(forKey: activeProfileDefaultsKey)
        }
        if let next = savedProfiles.sorted(by: { $0.lastUsedAt > $1.lastUsedAt }).first {
            self.activeProfileID = next.id
            UserDefaults.standard.set(next.id, forKey: activeProfileDefaultsKey)
            sessionState = .configured(next.serverURL)
        } else {
            sessionState = .unconfigured
        }
        connectionStatus = .offline
        serverConfig = nil
        serverConfigError = nil
        threadList = ThreadListStore()
    }

    private func startClient(serverURL: URL, bearerToken: String) async {
        if let existing = client {
            await existing.stop()
        }
        let conn = T3Connection(config: .init(serverURL: serverURL, bearerToken: bearerToken))
        let cli = T3Client(connection: conn)
        connection = conn
        client = cli

        let env = self
        await cli.addStatusListener { status in
            Task { @MainActor in
                switch status {
                case .offline:    env.connectionStatus = .offline
                case .connecting: env.connectionStatus = .connecting
                case .connected:  env.connectionStatus = .connected
                case .error(let message): env.connectionStatus = .error(message)
                }
            }
        }

        let connected = await cli.start()
        if connected {
            if let activeProfileID {
                markProfileUsed(activeProfileID)
            }
            await refreshServerConfig()
            await threadList.start(client: cli)
        }
    }

    func refreshServerConfig() async {
        guard let client else { return }
        do {
            let config = try await client.getServerConfig()
            await MainActor.run {
                self.serverConfig = config
                self.serverConfigError = nil
            }
        } catch {
            await MainActor.run {
                self.serverConfigError = error.localizedDescription
            }
        }
    }

    private var activeProfile: SavedServerProfile? {
        guard let activeProfileID else { return nil }
        return savedProfiles.first(where: { $0.id == activeProfileID })
    }

    private func tokenAccount(for profileID: SavedServerProfile.ID) -> String {
        tokenPrefix + profileID
    }

    private func upsertProfile(serverURL: URL, name: String) -> SavedServerProfile.ID {
        let now = Date()
        if let index = savedProfiles.firstIndex(where: { $0.serverURL.absoluteString == serverURL.absoluteString }) {
            savedProfiles[index].name = name
            savedProfiles[index].lastUsedAt = now
            persistProfiles()
            return savedProfiles[index].id
        }
        let profile = SavedServerProfile(
            id: UUID().uuidString,
            name: name,
            serverURL: serverURL,
            createdAt: now,
            lastUsedAt: now
        )
        savedProfiles.append(profile)
        persistProfiles()
        return profile.id
    }

    private func markProfileUsed(_ id: SavedServerProfile.ID) {
        guard let index = savedProfiles.firstIndex(where: { $0.id == id }) else { return }
        savedProfiles[index].lastUsedAt = Date()
        persistProfiles()
    }

    private func loadProfiles() {
        let defaults = UserDefaults.standard
        if let data = defaults.data(forKey: profilesDefaultsKey),
           let decoded = try? JSONDecoder().decode([SavedServerProfile].self, from: data) {
            savedProfiles = decoded
        }
        activeProfileID = defaults.string(forKey: activeProfileDefaultsKey)
    }

    private func persistProfiles() {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        if let data = try? encoder.encode(savedProfiles) {
            UserDefaults.standard.set(data, forKey: profilesDefaultsKey)
        }
    }

    private func migrateLegacySessionIfNeeded() {
        guard savedProfiles.isEmpty,
              let rawURL = KeychainStore.read(.serverURL),
              let url = URL(string: rawURL),
              let token = KeychainStore.read(.bearerToken) else {
            return
        }
        let profile = SavedServerProfile(
            id: UUID().uuidString,
            name: url.host ?? url.absoluteString,
            serverURL: url,
            createdAt: Date(),
            lastUsedAt: Date()
        )
        savedProfiles = [profile]
        activeProfileID = profile.id
        KeychainStore.save(token, account: tokenAccount(for: profile.id))
        KeychainStore.delete(.bearerToken)
        KeychainStore.delete(.serverURL)
        UserDefaults.standard.set(profile.id, forKey: activeProfileDefaultsKey)
        persistProfiles()
    }

    private static func bootstrapState() -> (savedProfiles: [SavedServerProfile], activeProfileID: SavedServerProfile.ID?, sessionState: SessionState) {
        let defaults = UserDefaults.standard
        let profilesKey = "savedServerProfiles"
        let activeKey = "activeServerProfileID"
        let tokenPrefix = "t3.bearer.token.profile."

        var profiles: [SavedServerProfile] = []
        if let data = defaults.data(forKey: profilesKey),
           let decoded = try? JSONDecoder().decode([SavedServerProfile].self, from: data) {
            profiles = decoded
        }
        var activeProfileID = defaults.string(forKey: activeKey)

        if profiles.isEmpty,
           let rawURL = KeychainStore.read(.serverURL),
           let url = URL(string: rawURL),
           let token = KeychainStore.read(.bearerToken) {
            let profile = SavedServerProfile(
                id: UUID().uuidString,
                name: url.host ?? url.absoluteString,
                serverURL: url,
                createdAt: Date(),
                lastUsedAt: Date()
            )
            profiles = [profile]
            activeProfileID = profile.id
            KeychainStore.save(token, account: tokenPrefix + profile.id)
            KeychainStore.delete(.bearerToken)
            KeychainStore.delete(.serverURL)
            defaults.set(profile.id, forKey: activeKey)
            let encoder = JSONEncoder()
            encoder.outputFormatting = [.sortedKeys]
            if let encodedProfiles = try? encoder.encode(profiles) {
                defaults.set(encodedProfiles, forKey: profilesKey)
            }
        }

        let activeProfile = profiles.first(where: { $0.id == activeProfileID })
            ?? profiles.sorted(by: { $0.lastUsedAt > $1.lastUsedAt }).first
        let sessionState: SessionState = activeProfile.map { .configured($0.serverURL) } ?? .unconfigured
        return (profiles, activeProfile?.id, sessionState)
    }
}

File: trifecta-android/app/src/main/java/com/belweave/trifecta/core/env/AppEnvironment.kt
package com.belweave.trifecta.core.env

import android.content.Context
import com.belweave.trifecta.core.auth.KeychainStore
import com.belweave.trifecta.core.models.ServerRuntimeConfig
import com.belweave.trifecta.core.networking.ConnectionState
import com.belweave.trifecta.core.networking.T3Client
import com.belweave.trifecta.core.networking.T3Connection
import com.belweave.trifecta.core.preferences.SavedProfileStore
import com.belweave.trifecta.core.preferences.SavedServerProfile
import com.belweave.trifecta.core.stores.ThreadListStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.net.URI
import java.time.Instant

/**
 * Singleton root that owns the active connection, client, profile list,
 * server config, and thread-list store. Mirrors the iOS `AppEnvironment`.
 */
class AppEnvironment(context: Context) {

    sealed class SessionState {
        object Unconfigured : SessionState()
        data class Configured(val serverURL: URI) : SessionState()
    }

    private val appContext = context.applicationContext
    private val keychain = KeychainStore(appContext)
    private val profileStore = SavedProfileStore(appContext)

    private val _sessionState = MutableStateFlow<SessionState>(SessionState.Unconfigured)
    val sessionState: StateFlow<SessionState> = _sessionState.asStateFlow()

    private val _connectionState = MutableStateFlow<ConnectionState>(ConnectionState.Offline)
    val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()

    private val _savedProfiles = MutableStateFlow<List<SavedServerProfile>>(emptyList())
    val savedProfiles: StateFlow<List<SavedServerProfile>> = _savedProfiles.asStateFlow()

    private val _activeProfileID = MutableStateFlow<String?>(null)
    val activeProfileID: StateFlow<String?> = _activeProfileID.asStateFlow()

    private val _serverConfig = MutableStateFlow<ServerRuntimeConfig?>(null)
    val serverConfig: StateFlow<ServerRuntimeConfig?> = _serverConfig.asStateFlow()

    private val _serverConfigError = MutableStateFlow<String?>(null)
    val serverConfigError: StateFlow<String?> = _serverConfigError.asStateFlow()

    val threadList = ThreadListStore()

    private val mutex = Mutex()
    private var connection: T3Connection? = null
    private var client: T3Client? = null
    private var statusJob: Job? = null

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    private val tokenPrefix = "t3.bearer.token.profile."

    suspend fun bootstrap() {
        val (profiles, activeID) = profileStore.load()
        _savedProfiles.value = profiles
        _activeProfileID.value = activeID
        val active = activeProfile() ?: profiles.maxByOrNull { it.lastUsedAt }
        if (active != null) {
            _sessionState.value = SessionState.Configured(active.serverURL)
        }
    }

    suspend fun resumeIfConfigured() {
        val active = activeProfile()
            ?: _savedProfiles.value.maxByOrNull { it.lastUsedAt }
            ?: return
        val token = keychain.read(tokenAccount(active.id)) ?: return
        _activeProfileID.value = active.id
        markProfileUsed(active.id)
        _sessionState.value = SessionState.Configured(active.serverURL)
        startClient(active.serverURL, token)
    }

    suspend fun configure(serverURL: URI, bearerToken: String, name: String? = null) {
        val displayName = (name?.trim()?.takeIf { it.isNotEmpty() })
            ?: (serverURL.host ?: serverURL.toString())
        val profileID = upsertProfile(serverURL, displayName)
        keychain.save(bearerToken, account = tokenAccount(profileID))
        _activeProfileID.value = profileID
        persistProfiles()
        _sessionState.value = SessionState.Configured(serverURL)
        startClient(serverURL, bearerToken)
    }

    suspend fun switchToProfile(id: String) {
        val profile = _savedProfiles.value.firstOrNull { it.id == id } ?: return
        val token = keychain.read(tokenAccount(id)) ?: return
        _activeProfileID.value = id
        markProfileUsed(id)
        _sessionState.value = SessionState.Configured(profile.serverURL)
        startClient(profile.serverURL, token)
    }

    suspend fun renameProfile(id: String, name: String) {
        val trimmed = name.trim()
        if (trimmed.isEmpty()) return
        _savedProfiles.value = _savedProfiles.value.map {
            if (it.id == id) it.copy(name = trimmed) else it
        }
        persistProfiles()
    }

    suspend fun removeProfile(id: String) {
        val wasActive = (_activeProfileID.value == id)
        keychain.delete(tokenAccount(id))
        _savedProfiles.value = _savedProfiles.value.filterNot { it.id == id }
        persistProfiles()
        if (!wasActive) return

        stopClient()
        threadList.reset()
        _serverConfig.value = null
        _serverConfigError.value = null
        _connectionState.value = ConnectionState.Offline

        val next = _savedProfiles.value.maxByOrNull { it.lastUsedAt }
        if (next != null) {
            _activeProfileID.value = next.id
            _sessionState.value = SessionState.Configured(next.serverURL)
            keychain.read(tokenAccount(next.id))?.let { token ->
                startClient(next.serverURL, token)
            }
        } else {
            _activeProfileID.value = null
            _sessionState.value = SessionState.Unconfigured
        }
    }

    suspend fun signOut() {
        stopClient()
        val active = _activeProfileID.value
        if (active != null) {
            keychain.delete(tokenAccount(active))
            _savedProfiles.value = _savedProfiles.value.filterNot { it.id == active }
            _activeProfileID.value = null
            persistProfiles()
        }
        val next = _savedProfiles.value.maxByOrNull { it.lastUsedAt }
        if (next != null) {
            _activeProfileID.value = next.id
            _sessionState.value = SessionState.Configured(next.serverURL)
        } else {
            _sessionState.value = SessionState.Unconfigured
        }
        _connectionState.value = ConnectionState.Offline
        _serverConfig.value = null
        _serverConfigError.value = null
        threadList.reset()
    }

    fun client(): T3Client? = client

    suspend fun refreshServerConfig() {
        val cli = client ?: return
        try {
            val cfg = cli.getServerConfig()
            _serverConfig.value = cfg
            _serverConfigError.value = null
        } catch (t: Throwable) {
            _serverConfigError.value = t.message
        }
    }

    private suspend fun startClient(serverURL: URI, bearerToken: String) {
        stopClient()
        val conn = T3Connection(T3Connection.Config(serverURL, bearerToken))
        val cli = T3Client(conn)
        mutex.withLock {
            connection = conn
            client = cli
        }

        statusJob = scope.launch {
            cli.addStatusListener { status ->
                _connectionState.value = when (status) {
                    is T3Connection.ConnectionStatus.Offline -> ConnectionState.Offline
                    is T3Connection.ConnectionStatus.Connecting -> ConnectionState.Connecting
                    is T3Connection.ConnectionStatus.Connected -> ConnectionState.Connected
                    is T3Connection.ConnectionStatus.Error -> ConnectionState.Error(status.message)
                }
            }
        }

        val connected = cli.start()
        if (connected) {
            _activeProfileID.value?.let { markProfileUsed(it) }
            refreshServerConfig()
            threadList.start(cli)
        }
    }

    private suspend fun stopClient() {
        statusJob?.cancel()
        statusJob = null
        threadList.stop()
        val (oldClient, oldConn) = mutex.withLock {
            val c = client
            val n = connection
            client = null
            connection = null
            c to n
        }
        oldClient?.stop()
        oldConn?.shutdown()
    }

    private fun activeProfile(): SavedServerProfile? {
        val id = _activeProfileID.value ?: return null
        return _savedProfiles.value.firstOrNull { it.id == id }
    }

    private fun tokenAccount(profileID: String): String = tokenPrefix + profileID

    private suspend fun upsertProfile(serverURL: URI, name: String): String {
        val now = Instant.now()
        val existingIndex = _savedProfiles.value.indexOfFirst {
            it.serverURLString == serverURL.toString()
        }
        return if (existingIndex >= 0) {
            val list = _savedProfiles.value.toMutableList()
            list[existingIndex] = list[existingIndex].copy(
                name = name,
                lastUsedAtIso = now.toString()
            )
            _savedProfiles.value = list
            list[existingIndex].id
        } else {
            val profile = SavedServerProfile.create(serverURL, name, now)
            _savedProfiles.value = _savedProfiles.value + profile
            profile.id
        }
    }

    private suspend fun markProfileUsed(id: String) {
        val now = Instant.now()
        _savedProfiles.value = _savedProfiles.value.map {
            if (it.id == id) it.copy(lastUsedAtIso = now.toString()) else it
        }
        persistProfiles()
    }

    private suspend fun persistProfiles() {
        profileStore.save(_savedProfiles.value, _activeProfileID.value)
    }

    companion object {
        @Volatile private var INSTANCE: AppEnvironment? = null

        fun get(context: Context): AppEnvironment {
            val existing = INSTANCE
            if (existing != null) return existing
            return synchronized(this) {
                val again = INSTANCE
                if (again != null) again
                else AppEnvironment(context).also { INSTANCE = it }
            }
        }
    }
}
